src/shapirowilk.jl

```diff
@@ -0,0 +1,250 @@
+
+let ROYSTON_COEFFS = Dict{String,Vector{Float64}}(
+        "C1" => [0.0E0, 0.221157E0, -0.147981E0, -0.207119E1, 0.4434685E1, -0.2706056E1],
+        "C2" => [0.0E0, 0.42981E-1, -0.293762E0, -0.1752461E1, 0.5682633E1, -0.3582633E1],
+        "C3" => [0.5440E0, -0.39978E0, 0.25054E-1, -0.6714E-3],
+        "C4" => [0.13822E1, -0.77857E0, 0.62767E-1, -0.20322E-2],
+        "C5" => [-0.15861E1, -0.31082E0, -0.83751E-1, 0.38915E-2],
+        "C6" => [-0.4803E0, -0.82676E-1, 0.30302E-2],
+        "C7" => [0.164E0, 0.533E0],
+        "C8" => [0.1736E0, 0.315E0],
+        "C9" => [0.256E0, -0.635E-2],
+        "G" => [-0.2273E1, 0.459E0]
+    )
+
+    for (s, c) in ROYSTON_COEFFS
+        @eval $(Symbol("__RS92_" * s))(x) = Base.Math.@horner(x, $(c...))
+    end
+end
+
+#=
+The following hardcoded constants has been replaced by more precise values:
+
+SQRTH = sqrt(2.0)/2.0 # 0.70711E0
+TH = 3/8 # 0.375E0
+SMALL = eps(1.0) # 1E-19
+PI6 = π/6 # 0.1909859E1
+STQR = asin(sqrt(0.75)) # 0.1047198E1
+=#
+
+struct SWCoeffs <: AbstractVector{Float64}
+    N::Int
+    A::Vector{Float64}
+end
+
+Base.size(SWc::SWCoeffs) = (SWc.N,)
+Base.IndexStyle(::Type{SWCoeffs}) = IndexLinear()
+
+function Base.getindex(SWc::SWCoeffs, i::Int)
+    if firstindex(SWc.A) ≤ i ≤ lastindex(SWc.A)
+        return @inbounds SWc.A[i]
+    elseif isodd(SWc.N) && i == lastindex(SWc.A) + 1
+        return zero(eltype(SWc))
+    else
+        return -SWc.A[SWc.N+1-i]
+    end
+end
+
+function SWCoeffs(N::Int)
+    if N < 3
+        throw(ArgumentError("N must be greater than or equal to 3: $N"))
+    elseif N == 3 # exact
+        return SWCoeffs(N, [sqrt(2.0) / 2.0])
+    else
+        # Weisberg&Bingham 1975 statistic; store only positive half of m:
+        # it is (anti-)symmetric; hence '2' factor below
+        m = [-quantile(Normal(), (i - 3 / 8) / (N + 1 / 4)) for i in 1:div(N, 2)]
+        mᵀm = 2sum(abs2, m)
+        x = 1 / sqrt(N)
+        a₁ = m[1] / sqrt(mᵀm) + __RS92_C1(x) # aₙ = cₙ + (...)
+        if N ≤ 5
+            ϕ = (mᵀm - 2m[1]^2) / (1 - 2a₁^2)
+            m .= m ./ sqrt(ϕ) # A, but reusing m to save allocs
+            m[1] = a₁
+        else
+            a₂ = m[2] / sqrt(mᵀm) + __RS92_C2(x) # aₙ₋₁ = cₙ₋₁ + (...)
+            ϕ = (mᵀm - 2m[1]^2 - 2m[2]^2) / (1 - 2a₁^2 - 2a₂^2)
+            m .= m ./ sqrt(ϕ) # A, but reusing m to save allocs
+            m[1], m[2] = a₁, a₂
+        end
+
+        return SWCoeffs(N, m)
+    end
+end
+
+function swstat(X::AbstractArray{<:Real}, A::SWCoeffs)
+    if last(X) - first(X) < length(X) * eps()
+        throw("data seems to be constant!")
+    end
+    AX = dot(A, X)
+    m = mean(X)
+    S² = sum(x -> abs2(x - m), X)
+    return AX^2 / S²
+end
+
+function pvalue(W::Real, A::SWCoeffs, N1=A.N)
+    A.N != N1 && throw("not implemented yet!")
+    if A.N == 3 # exact by Shapiro&Wilk 1965
+        return π / 6 * (asin(sqrt(W)) - asin(sqrt(0.75)))
+    elseif A.N ≤ 11
+        γ = __RS92_G(A.N)
+        if γ ≤ log(1 - W)
```

Comment:
```suggestion
        if γ ≤ log1p(-W)
```
This is more accurate if `W` is small and negative. There are a couple of other places where this suggestion could be applied as well.